### PR TITLE
Cross off a few tslibs-TODOs

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is a hodge-podge addressing a few of the items in #17652.  Individual commits should be item-specific(ish).

- remove unused lib.arrmap
- de-privatize _checknull_with_nat
- fix a couple of C warnings caused by timedelta_struct
- remove Period._comparables

### CODE REVIEW

File: pandas/_libs/src/datetime/np_datetime.c
Comment (WillAyd) [inline pandas/_libs/src/datetime/np_datetime.c]: Thanks for fixing this typo. Do we even need the assignment at all though? Worth considering removal?

**Comment by WillAyd** [comment]
While you are touching this file I just noticed a typo in the below line:

https://github.com/jbrockmendel/pandas/blob/6ef9227fda1ee595eecf298e65ba99d6f9180db2/pandas/_libs/src/datetime/np_datetime.c#L1087

The word datetime should read timedelta (copy / paste issue). Figured it’s worth fixing while you are touching the file but if you wanted me to fix in a separate issue and PR let me know

File: pandas/_libs/src/datetime/np_datetime.c
Comment (jbrockmendel) [reply pandas/_libs/src/datetime/np_datetime.c]: I'm advocating moving the timedelta_struct stuff out of src/ entirely.  The src/ directory is a PITA and we can put it in cython directly without taking a performance hit.

**Comment by jbrockmendel** [comment]
Good call. There’s also some indentation in timedeltastruct I’ll get to, could use a double-checking to make sure I don’t screw it up.

**Comment by WillAyd** [comment]
ping me when complete and I’d be glad to double check

**Comment by jbrockmendel** [comment]
@WillAyd ping.  Should there be an extra indent after `case PANDAS_FR_ns:`?  I'm not that familiar with C syntax.

If we do move the timedelta_struct stuff into cython, let's not bother with the conditioning on meta.base since only one is ever used.

File: pandas/_libs/src/datetime/np_datetime.c
Comment (jreback) [inline pandas/_libs/src/datetime/np_datetime.c]: I *think* we already lint this file (cpplint). is this not picked up? rather have the proper config settings than manually changing things.

File: pandas/_libs/tslib.pyx
Comment (jreback) [inline pandas/_libs/tslib.pyx]: these array_* should be moved to conversion I think (obviously new PR), then tslib.pyx is moot?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/datetime/np_datetime.c
Comment (jbrockmendel) [reply pandas/_libs/src/datetime/np_datetime.c]: No argument here, but cpplint is well outside my domain of competence.  I only noticed this indentation when trying to move this func into cython.

Comment (jreback) [reply pandas/_libs/src/datetime/np_datetime.c]: @WillAyd any experience with cpplint? want to give this a shot, e.g. making sure we are linting the *.c (more/better)?

File: pandas/_libs/tslib.pyx
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: Let's hold off on that.  I'd like to do auditing before moving, and the array_* funcs may need some work/thought (#17697)

**Comment by jbrockmendel** [comment]
Travis error in test_geopandas

```
>   from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
E   ImportError: libpoppler.so.66: cannot open shared object file: No such file or directory
../../../miniconda3/envs/pandas/lib/python3.6/site-packages/fiona/collection.py:9: ImportError
```

**Comment by jreback** [comment]
no idea, maybe something is in process of being updated

File: pandas/_libs/src/datetime/np_datetime.c
Comment (WillAyd) [reply pandas/_libs/src/datetime/np_datetime.c]: Can't claim experience with cpplint but looking at the bash script that's it appears to already be hitting all of the files in this folder. I ran both versions through cpplint locally and neither version caused any issue with the highest level of verbosity and neither raised an alarm. Surprisingly I couldn't get an error when changing the line to a 1 or 3 space indent either, so I think there might be some bug with cpplint itself.

In looking at Google's style guide they suggest a 2 space indent. To match that and keep consistent with the rest of the file I would actually suggest removing the two spaces added here and remove the two from the indented block after the ``else`` statement instead

https://google.github.io/styleguide/cppguide.html#Conditionals

Comment (jreback) [reply pandas/_libs/src/datetime/np_datetime.c]: see https://github.com/apache/arrow?files=1

for some hints in how to lint / enforce this as well

**Comment by WillAyd** [comment]
Referring to the [Google style guide](https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements) again we should have a 2 space indent after the ``case PANDAS_FR_ns:``. 

As I mentioned on a previous comment, I have a feeling that there's either a bug, missing configuration for cpplint to detect those whitespace issues. I'm not overly familiar with the tool so need to research further and advise back.

File: pandas/_libs/src/datetime/np_datetime.c
Comment (WillAyd) [reply pandas/_libs/src/datetime/np_datetime.c]: Are you saying that within arrow cpplint is calling out those whitespace issues? I tried replicating the filters I saw in that project via:

    cpplint --verbose=2 --filter=-whitespace/comments,-readability/todo,-build/header_guard,-build/c++11,-runtime/references,-build/include_order pandas/_libs/src/datetime/np_datetime.c

And then tried an even more verbose version without any filter
    
    cpplint --verbose=1 pandas/_libs/src/datetime/np_datetime.c

And neither made any mention of those lines, whether I set them at 2, 4 or even 3 space indents.

Comment (jreback) [reply pandas/_libs/src/datetime/np_datetime.c]: no just a guess that they have more options set for linting

Comment (WillAyd) [reply pandas/_libs/src/datetime/np_datetime.c]: Looking at the cpplint source, I don't see anything that checks for proper 2-space indentation of conditional statements, nor was there any documentation hinting that it could. The ``whitespace/indentat`` check offered only seems to make sure that you don't start with 1 or 3 spaces, but doesn't do much beyond that. Unfortunately I don't see any programmatic options to continually validate this with that tool as is

**Comment by jbrockmendel** [comment]
@WillAyd I'm way out of my element w/r/t C linting, will have to trust you that the 2-spaces is correct.  Are my eye playing tricks on me when the indentation looks inconsistent?  i.e. should I just revert all the indentation edits?

Isn't the switch/case usage here unnecessary anyway?  Or are there plans to implement and use other units?

**Comment by WillAyd** [comment]
You are correct that the switch/case usage is at the moment unnecessary, because the method is only being called with nanosecond frequency. The reason I kept it in there was to mirror the implementation that was done for datetime.

As far as the indentation is concerned, that's up to @jreback. I would suggest going forward with your 4 space indentation because it's consistent with the rest of the file. From what I can tell cpplint isn't going to care either way.

**Comment by jbrockmendel** [comment]
> The reason I kept it in there was to mirror the implementation that was done for datetime.

That makes sense.  In the interest of keeping this already-too-wide-ranging PR from expanding scope, I'd like to leave it as is for now, i.e. use 4-space consistently and keep the unnecessary switch/case.  Later I think we should implement the pandas_timedeltastruct and td64_to_tdstruct in tslibs.np_timedeltas directly in cython.  If there isn't a consensus on the latter, achieving it on the former can be good enough for now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel as always, keeping scope narrow is better per PR

**Comment by jreback** [comment]
I am not ticking off boxes in #17652 but pls do (and link to the issues)

**Pull request merged**
